lib/statsd/instrument/strict.rb

```diff
@@ -0,0 +1,84 @@
+    #
+    # - It will make sure the metric methods are not retuning a Metric instance.
+    # - It will only accept
+    #
+    # You can enable thois monkeypatch by changing your Gemfile as follows:
+    #
+    #     gem 'statsd-instrumenr', require: 'statsd/instrument/strict'
+    #
+    # By doing this as part of your QA/CI, you can find where you are still using deprecated patterns,
+    # and fix them before the deprecated behavior is removed in the next major version.
+    #
+    # This monkeypatch is not meant to be used in production.
+    module Strict
+      def increment(key, value = 1, sample_rate: nil, tags: nil, prefix: StatsD.prefix, no_prefix: false)
+        raise ArgumentError, "StatsD.increment does not accept a block" if block_given?
+        raise ArgumentError, "The value argument should be an integer" unless value.is_a?(Integer)
+        raise ArgumentError, "The sample_rate argument should be a number" unless sample_rate.nil? || sample_rate.is_a?(Numeric)
+        raise ArgumentError, "The tags argument should be a hash or an array" unless tags.nil? || tags.is_a?(Hash) || tags.is_a?(Array)
+        super
+      end
+
+      def gauge(key, value, sample_rate: nil, tags: nil, prefix: StatsD.prefix, no_prefix: false)
+        raise ArgumentError, "StatsD.gauge does not accept a block" if block_given?
+        raise ArgumentError, "The value argument should be an number" unless value.is_a?(Numeric)
+        raise ArgumentError, "The sample_rate argument should be a number" unless sample_rate.nil? || sample_rate.is_a?(Numeric)
+        raise ArgumentError, "The tags argument should be a hash or an array" unless tags.nil? || tags.is_a?(Hash) || tags.is_a?(Array)
+        super
+      end
+
+      def histogram(key, value, sample_rate: nil, tags: nil, prefix: StatsD.prefix, no_prefix: false)
+        raise ArgumentError, "StatsD.histogram does not accept a block" if block_given?
+        raise ArgumentError, "The value argument should be an number" unless value.is_a?(Numeric)
+        raise ArgumentError, "The sample_rate argument should be a number" unless sample_rate.nil? || sample_rate.is_a?(Numeric)
+        raise ArgumentError, "The tags argument should be a hash or an array" unless tags.nil? || tags.is_a?(Hash) || tags.is_a?(Array)
+        super
+      end
+
+      def set(key, value, sample_rate: nil, tags: nil, prefix: StatsD.prefix, no_prefix: false)
+        raise ArgumentError, "The sample_rate argument should be a number" unless sample_rate.nil? || sample_rate.is_a?(Numeric)
+        raise ArgumentError, "The tags argument should be a hash or an array" unless tags.nil? || tags.is_a?(Hash) || tags.is_a?(Array)
+        super
+      end
+
+      def measure(key, value = nil, sample_rate: nil, tags: nil, prefix: StatsD.prefix, no_prefix: false, as_dist: false, &block)
+        if block_given?
+          raise ArgumentError, "The value argument should be set to nil when providing a block" unless value.nil?
```

Comment:
From the library user's perspective, they should just think that they value shouldn't be given when calling it with a block.

```suggestion
          raise ArgumentError, "A value argument should be given when providing a block" unless value.nil?
```